heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/metricscache/MetricsCacheQueryUtils.java

```diff
@@ -0,0 +1,290 @@
+      // only one component
+      request1.componentNameInstanceId.put(componentName, instances);
+      // if there are instances specified
+      instances.addAll(request.getInstanceIdList());
+    }
+
+    Set<String> metrics = new HashSet<>();
+    request1.metricNames = metrics;
+    if (request.getMetricCount() > 0) {
+      metrics.addAll(request.getMetricList());
+    } // empty list means no metrics
+
+    // default: the whole time horizon
+    request1.startTime = 0;
+    request1.endTime = Long.MAX_VALUE;
+    if (request.hasInterval()) { // endTime = now
+      request1.endTime = System.currentTimeMillis();
+
+      long interval = request.getInterval(); // in seconds
+      if (interval <= 0) { // means all
+        request1.startTime = 0;
+      } else { // means [-interval, now]
+        request1.startTime = request1.endTime - interval * 1000;
+      }
+    } else {
+      request1.startTime = request.getExplicitInterval().getStart() * 1000;
+      request1.endTime = request.getExplicitInterval().getEnd() * 1000;
+    }
+
+    // default: aggregate all metrics
+    request1.minutely = 0;
+    if (request.hasMinutely()) {
+      request1.minutely = request.getMinutely() ? 1 : 0;
+    }
+
+    return request1;
+  }
+
+  /**
+   * compatible with com.twitter.heron.proto.tmaster.TopologyMaster.MetricResponse
+   */
+  public static TopologyMaster.MetricResponse Convert(MetricResponse response,
+                                                      MetricRequest request) {
+    TopologyMaster.MetricResponse.Builder builder =
+        TopologyMaster.MetricResponse.newBuilder();
+    builder.setInterval((request.endTime - request.startTime) / 1000); // in seconds
+
+    // default OK if we have response to build already
+    builder.setStatus(Common.Status.newBuilder().setStatus(Common.StatusCode.OK));
+
+    // instanceId -> [metricName -> metricValue]
+    // componentName is ignored, since there is only one component in the query
+    Map<String, Map<String, List<MetricTimeRangeValue>>> aggregation =
+        new HashMap<>();
+    for (MetricDatum datum : response.metricList) {
+      String instanceId = datum.instanceId;
+      String metricName = datum.metricName;
+      List<MetricTimeRangeValue> metricValue = datum.metricValue;
+      // prepare
+      if (!aggregation.containsKey(instanceId)) {
+        aggregation.put(instanceId, new HashMap<String, List<MetricTimeRangeValue>>());
+      }
+      if (!aggregation.get(instanceId).containsKey(metricName)) {
+        aggregation.get(instanceId).put(metricName, new ArrayList<MetricTimeRangeValue>());
+      }
+      // aggregate
+      aggregation.get(instanceId).get(metricName).addAll(metricValue);
+    }
+
+    // add TaskMetric
+    for (String instanceId : aggregation.keySet()) {
+      TopologyMaster.MetricResponse.TaskMetric.Builder builder1 =
+          TopologyMaster.MetricResponse.TaskMetric.newBuilder();
+
+      builder1.setInstanceId(instanceId);
```

Comment:
Please use better naming than builder[1|2|3] in this class (e.g. metricResponseBuilder, taskMetricBuilder, indivudualMetricBuilder). Better to be more verbose than to be cryptic. The code will read better.